qiskit/algorithms/eigensolvers/vqd.py

```diff
@@ -0,0 +1,545 @@
+            expectation_values = estimator_job.result().values
+        except Exception as exc:
+            raise AlgorithmError("The primitive job failed!") from exc
+
+        std_devs = self._compute_std_devs(estimator_job, len(expectation_values))
+
+        # Discard values below threshold
+        observables_means = expectation_values * (np.abs(expectation_values) > threshold)
+        # zip means and standard deviations into tuples
+        observables_results = list(zip(observables_means, std_devs))
+
+        # Return None eigenvalues for None operators if observables is a list.
+        return self._prepare_result(observables_results, observables)
+
+    @staticmethod
+    def _prepare_result(
+        observables_results: List[Tuple[complex, complex]],
+        observables: ListOrDict[BaseOperator | PauliSumOp],
+    ) -> ListOrDict[Tuple[complex, complex]]:
+        """
+        Prepares a list of eigenvalues and standard deviations from ``observables_results`` and
+        ``observables``.
+
+        Args:
+            observables_results: A list of of tuples (mean, standard deviation).
+            observables: A list or a dictionary of operators whose expectation values are to be
+                calculated.
+
+        Returns:
+            A list or a dictionary of tuples (mean, standard deviation).
+        """
+
+        if isinstance(observables, list):
+            observables_eigenvalues = [None] * len(observables)
+            key_value_iterator = enumerate(observables_results)
+        else:
+            observables_eigenvalues = {}
+            key_value_iterator = zip(observables.keys(), observables_results)
+
+        for key, value in key_value_iterator:
+            if observables[key] is not None:
+                observables_eigenvalues[key] = value
+        return observables_eigenvalues
+
+    @staticmethod
+    def _compute_std_devs(
+        estimator_result: EstimatorResult,
+        results_length: int,
+    ) -> List[complex | None]:
+        """
+        Calculates a list of standard deviations from expectation values of observables provided. If
+        the choice of an underlying hardware is not shot-based and hence does not provide variance data,
+        the standard deviation values will be set to ``None``.
+
+        Args:
+            estimator_result: An estimator result.
+            results_length: Number of expectation values calculated.
+
+        Returns:
+            A list of standard deviations.
+        """
+        if not estimator_result.metadata:
+            return [0] * results_length
+
+        std_devs = []
+        for metadata in estimator_result.metadata:
+            if metadata and "variance" in metadata.keys() and "shots" in metadata.keys():
+                variance = metadata["variance"]
+                shots = metadata["shots"]
+                std_devs.append(np.sqrt(variance / shots))
+            else:
+                std_devs.append(None)
+
+        return std_devs
```

Comment:
These 3 methods will be removed once #8683 is merged, and replaced by an import of `estimate_observables()`